File: src/za/co/adhd_developers/questions/Question14.java
package za.co.adhd_developers.questions;

import java.util.Hashtable;

import za.co.adhd_developers.tools.Question;

/**
 * Created by Grant on 2017/07/29.
 */
public class Question14 implements Question
{
    int answer = 0;
    long chain = 0;
    Hashtable<Integer, Long> chains = new Hashtable<>();

    private long calculateEven(long term)
    {
        return term/2;
    }

    private long calculateOdd(long term)
    {
        return (term * 3) + 1;
    }

    @Override
    public void doWork()
    {
        for (int i = 1; i < 1000000; i++)
        {
            long num = (long) i;
            long count = 0;

            while (num != 0)
            {
                if (this.chains.keySet().contains(num))
                {
                    long otherCount = this.chains.get(num);

                    count = count + otherCount;

                    this.chains.put(i, count);
                    num = 0;
                }
                else if (num == 1)
                {
                    count++;
                    this.chains.put(i, count);
                    num = 0;
                }
                else
                {
                    count++;
                    if (num % 2 == 0)
                    {
                        num = calculateEven(num);
                    }
                    else
                    {
                        num = calculateOdd(num);
                    }
                }
            }
        }

        for (Integer key : this.chains.keySet())
        {
            long tmp_chain = this.chains.get(key);

            if (tmp_chain > this.chain)
            {
                this.chain = tmp_chain;
                this.answer = key;
            }
        }
    }

    @Override
    public void printAnswer()
    {
        System.out.println("=================================");
        System.out.println("Question 14");
//        System.out.println("=================================");
//        System.out.println(this.chains.size());
        System.out.println("The starting number with the longest chain is: " + this.answer + " (" + this.chain + ")");
        System.out.println("=================================");
    }
}

File: src/za/co/adhd_developers/kotlin/questions/Question14.kt
package za.co.adhd_developers.kotlin.questions

import za.co.adhd_developers.kotlin.Question
import java.util.*

class Question14 : Question {

    var answer = 0L
    var chain = 0

    override fun doWork() {

        var results = Hashtable<Long, Int>()

        for (i in 1L..999999L) {
            var number = i
            var count = 0
            while (number > 1) {

                if (results.containsKey(number)) {
                    count += results.getValue(number)
                    number = 1
                } else {
                    number = if (number.rem(2) == 0L) {
                        number.div(2L)
                    } else {
                        number.times(3L).plus(1L)
                    }
                    count++
                }
            }
            results.put(i, count)
        }

        this.chain = results.values.sortedDescending().first()

        results.filter { entry -> entry.value == this.chain }.forEach { t, u -> this.answer = t }
    }

    override fun printAnswer() {
        println("=================================")
        println("Question 14")
        println("The starting number with the longest chain is: " + this.answer + " (" + this.chain + ")")
        println("=================================")
    }
}